cli/kafka/runner.go

```diff
@@ -7,61 +7,103 @@ import (
 	fbbatch "github.com/featurebasedb/featurebase/v3/batch"
 	"github.com/featurebasedb/featurebase/v3/errors"
 	"github.com/featurebasedb/featurebase/v3/idk"
-	"github.com/featurebasedb/featurebase/v3/idk/kafka_static"
+	"github.com/featurebasedb/featurebase/v3/idk/common"
+	"github.com/featurebasedb/featurebase/v3/idk/kafka"
+	"github.com/featurebasedb/featurebase/v3/idk/kafka_sasl"
 	"github.com/featurebasedb/featurebase/v3/logger"
 )
 
 // Runner is a CLI-specific kafka consumer. It's similar to
 // idk.kafka_static.Main in that it embeds idk.Main and contains additional
 // functionality specific to its use case.
 type Runner struct {
-	idk.Main   `flag:"!embed"`
-	KafkaHosts []string       `help:"Comma separated list of host:port pairs for Kafka."`
-	Group      string         `help:"Kafka group."`
-	Topics     []string       `help:"Kafka topics to read from."`
-	Timeout    time.Duration  `help:"Time to wait for more records from Kafka before flushing a batch. 0 to disable."`
-	Header     []idk.RawField `help:"Header configuration."`
+	idk.Main             `flag:"!embed"`
+	idk.ConfluentCommand `flag:"!embed"`
+	Hosts                []string       `help:"Comma separated list of host:port pairs for Kafka."`
+	Group                string         `help:"Kafka group."`
+	Topics               []string       `help:"Kafka topics to read from."`
+	Timeout              time.Duration  `help:"Time to wait for more records from Kafka before flushing a batch. 0 to disable."`
+	Header               []idk.RawField `help:"Header configuration."`
 }
 
+// Configure and return *runner with common elements of all kafka runners
 func NewRunner(cfg ConfigForIDK, batcher fbbatch.Batcher, logWriter io.Writer) *Runner {
 	idkMain := idk.NewMain()
 	idkMain.IDField = cfg.IDField
+	idkMain.PrimaryKeyFields = cfg.PrimaryKeys
 	idkMain.Index = cfg.Table
 	idkMain.Batcher = batcher
 	idkMain.BatchSize = cfg.BatchSize
 	idkMain.BatchMaxStaleness = cfg.BatchMaxStaleness
+	idkMain.MaxMsgs = uint64(cfg.MaxMessages)
 	idkMain.SetBasic()
 	idkMain.SetLog(logger.NewStandardLogger(logWriter))
-
+	idkMain.OffsetMode = true
+	idkMain.Namespace = "sql_kafka_runner"
+	idkMain.Pprof = "" // don't initialize pprof until we actually use it in tests
 	kr := &Runner{
-		Main:       *idkMain,
-		KafkaHosts: cfg.Hosts,
-		Group:      cfg.Group,
-		Topics:     cfg.Topics,
-		Header:     cfg.Fields,
-		Timeout:    cfg.Timeout,
+		Main:    *idkMain,
+		Hosts:   cfg.Hosts,
+		Group:   cfg.Group,
+		Topics:  cfg.Topics,
+		Header:  cfg.Fields,
+		Timeout: cfg.Timeout,
+	}
+
+	// NewSource should be set based on the encoding of the source (e.g. JSON, Avro)
+	if cfg.Encode == encodingTypeAvro {
+		kr.GetAvroNewSource(cfg)
+	} else if cfg.Encode == encodingTypeJSON {
+		kr.GetJSONNewSource(cfg)
```

Comment:
If it's not one of these two options, we'll catch it in the call to `ValidateConfig()`. Also, the default config options is "json" so if the user didn't specify a `encode` config, it would be "json".

But, I'll make this change so it's more readable and we have multiple checks for the same thing. Just need to update a few places if we ever wanted to make a change to the defaults. But I don't see that as an issue.